esmvalcore/_data_finder.py

```diff
@@ -231,65 +284,104 @@ def _find_input_dirs(variable, rootpath, drs):
     return (files, dirnames, filenames)
 
 
 def get_output_file(variable, preproc_dir):
     """Return the full path to the output (preprocessed) file."""
-    cfg = get_project_config(variable['project'])
+    cfg = get_project_config(variable["project"])
 
     # Join different experiment names
-    if isinstance(variable.get('exp'), (list, tuple)):
+    if isinstance(variable.get("exp"), (list, tuple)):
         variable = dict(variable)
-        variable['exp'] = '-'.join(variable['exp'])
+        variable["exp"] = "-".join(variable["exp"])
 
     outfile = os.path.join(
         preproc_dir,
-        variable['diagnostic'],
-        variable['variable_group'],
-        _replace_tags(cfg['output_file'], variable)[0],
+        variable["diagnostic"],
+        variable["variable_group"],
+        _replace_tags(cfg["output_file"], variable)[0],
     )
-    if variable['frequency'] != 'fx':
-        outfile += '_{start_year}-{end_year}'.format(**variable)
-    outfile += '.nc'
+    if variable["frequency"] != "fx":
+        outfile += "_{start_year}-{end_year}".format(**variable)
+    outfile += ".nc"
     return outfile
 
 
 def get_statistic_output_file(variable, preproc_dir):
     """Get multi model statistic filename depending on settings."""
     template = os.path.join(
         preproc_dir,
-        '{diagnostic}',
-        '{variable_group}',
-        '{dataset}_{mip}_{short_name}_{start_year}-{end_year}.nc',
+        "{diagnostic}",
+        "{variable_group}",
+        "{dataset}_{mip}_{short_name}_{start_year}-{end_year}.nc",
     )
 
     outfile = template.format(**variable)
 
     return outfile
+
+
+def dir_to_var(dirname, basepath, project, drs):
+    """Convert directory path to variable :obj:`dict`."""
+    if dirname != os.sep:
+        dirname = dirname.rstrip(os.sep)
+    if basepath != os.sep:
+        basepath = basepath.rstrip(os.sep)
+    path_template = _select_drs("input_dir", drs, project).rstrip(os.sep)
+    rel_dir = os.path.relpath(dirname, basepath)
+    keys = path_template.split(os.sep)
+    vals = rel_dir.split(os.sep)
+    if len(keys) != len(vals):
+        raise ValueError(
+            f"Cannot extract tags '{path_template}' from directory "
+            f"'{rel_dir}' (root: '{basepath}') with different numbers of "
+            f"elements")
+    variable = {}
+    for (idx, full_key) in enumerate(keys):
+        matches = re.findall(r'.*\{(.*)\}.*', full_key)
+        if len(matches) != 1:
+            continue
+        key = matches[0]
+        regex = rf"{full_key.replace(key, '(.*)')}"
+        regex = regex.replace('{', '').replace('}', '')
+        matches = re.findall(regex, vals[idx])
+        while '' in matches:
+            matches.remove('')
+        if len(matches) != 1:
+            raise ValueError(
+                f"Regex pattern '{regex}' for '{full_key}' cannot be "
+                f"(uniquely) matched to element '{vals[idx]}' in directory "
+                f"'{dirname}'")
```

Comment:
this is a brilliantly opaque error message, can you pls make it more user-friendly? Imagine you're a first time user and this pops up :laughing: